pytensor/tensor/subtensor.py

```diff
@@ -2585,26 +2581,49 @@ def R_op(self, inputs, eval_points):
         return self.make_node(eval_points[0], *inputs[1:]).outputs
 
     def infer_shape(self, fgraph, node, ishapes):
-        indices = node.inputs[1:]
-        index_shapes = list(ishapes[1:])
-        for i, idx in enumerate(indices):
-            if (
+        def is_bool_index(idx):
+            return (
                 isinstance(idx, (np.bool_, bool))
                 or getattr(idx, "dtype", None) == "bool"
-            ):
-                raise ShapeError(
-                    "Shape inference for boolean indices is not implemented"
+            )
+
+        indices = node.inputs[1:]
+        index_shapes = []
+        for idx, ishape in zip(indices, ishapes[1:]):
+            # Mixed bool indexes are converted to nonzero entries
+            # Note: We don't assert that the original shape is valid!
+            if is_bool_index(idx):
+                index_shapes.extend(
+                    (shape_i(nz_dim, 0, fgraph=fgraph),) for nz_dim in nonzero(idx)
                 )
             # The `ishapes` entries for `SliceType`s will be None, and
             # we need to give `indexed_result_shape` the actual slices.
-            if isinstance(getattr(idx, "type", None), SliceType):
-                index_shapes[i] = idx
+            elif isinstance(getattr(idx, "type", None), SliceType):
+                index_shapes.append(idx)
+            else:
+                index_shapes.append(ishape)
 
-        res_shape = indexed_result_shape(
-            ishapes[0], index_shapes, indices_are_shapes=True
+        res_shape = list(
+            indexed_result_shape(ishapes[0], index_shapes, indices_are_shapes=True)
         )
+
+        adv_indices = [idx for idx in indices if not is_basic_idx(idx)]
+        bool_indices = [idx for idx in adv_indices if is_bool_index(idx)]
+
+        # Special logic when the only advanced index group is of bool type.
+        # We can replace the nonzeros by a sum of the whole bool variable.
+        # Note: We don't assert that the original shape is valid!
```

Comment:
I think this is consistent with other `infer_shape` implementations. For example the `Elemwise.infer_shape` does not assert the inputs can broadcast correctly.

This is explicitly documented: https://pytensor.readthedocs.io/en/latest/tutorial/shape_info.html#problems-with-shape-inference

Reply:
Okay maybe more relevant to justify this decision, invalid advanced integer broadcasting also does not raise:
```python
import pytensor.tensor as pt
x = pt.vector("x")
x[[0, 20, 0]].shape.eval({x: [0, 1, 2]})  # [3]
```

Reply:
Okay this seems to be a bit out of the scope of this one PR, so I opened a separate issue here: https://github.com/brandonwillard/aesara/commit/064e72f4c37d069dbb5e829e38cf196d32c89fa8